job_creator.php

```diff
@@ -0,0 +1,270 @@
+                    $matrix['include'][] = $this->createJob(0, [
+                        'phpcoverage' => true
+                    ]);
+                } else {
+                    $matrix['include'][] = $this->createJob(2, [
+                        'db' => DB_MYSQL_57_PDO,
+                        'phpcoverage' => true
+                    ]);
+                }
+            }
+            // endtoend / behat
+            if ($run['endtoend'] && file_exists('behat.yml')) {
+                $matrix['include'][] = $this->createJob(0, [
+                    'endtoend' => true,
+                    'endtoend_suite' => 'root'
+                ]);
+                if (!$simpleMatrix) {
+                    $matrix['include'][] = $this->createJob(3, [
+                        'db' => DB_MYSQL_80,
+                        'endtoend' => true,
+                        'endtoend_suite' => 'root'
+                    ]);
+                }
+            }
+            // javascript tests
+            if (file_exists('package.json') && $run['js']) {
+                $matrix['include'][] = $this->createJob(0, [
+                    'js' => true
+                ]);
+            }
+        }
+
+        // extra jobs
+        foreach ($extraJobs as $arr) {
+            $matrix['include'][] = $this->createJob(0, $arr);
+        }
+
+        // convert everything to strings and sanatise values
+        foreach ($matrix['include'] as $i => $job) {
+            foreach ($job as $key => $val) {
+                if ($val === true) {
+                    $val = 'true';
+                }
+                if ($val === false) {
+                    $val = 'false';
+                }
+                // all values must be strings
+                $val = (string) $val;
+                // remove any dodgy characters
+                $val = str_replace(["\r", "\n", "\t", "'", '"', '&', '|'], '', $val);
+                // ascii chars only - https://www.regular-expressions.info/posixbrackets.html
+                $val = preg_replace('#[^\x20-\x7E]#', '', $val);
+                // limit name_suffix length and be strict as it is used in the artifact file name
+                if ($key === 'name_suffix') {
+                    if (strlen($val) > 20) {
+                        $val = preg_replace('#[^a-zA-Z0-9_\- ]#', '', $val);
+                        $val = substr($val, 0, 20);
+                    }
+                }
+                // be strict with composer_require_extra
+                if ($key === 'composer_require_extra') {
+                    $val = preg_replace('#[^A-Za-z0-9\-\.\^\/~: ]#', '', $val);
+                }
+                // add value back to matrix
+                $matrix['include'][$i][$key] = $val;
+            }
+        }
+
+        // job/artifacts names
+        foreach ($matrix['include'] as $i => $job) {
+            $name = [
+                $job['php']
+            ];
+            if ($job['composer_args'] == '--prefer-lowest') {
```

Comment:
Is it possible for this to be a string with multiple args? e.g. `'--prefer-lowest --prefer-source -q'`? It looks like that's no currently possible, but is set up to allow it in the future. Might be best to check if this is in the string instead. Something like:
```php
if (preg_match('/--prefer-lowest\b/', $job['composer_args'])) {
```

Reply:
composer_args is only set in this php script, it's no an input. If change things later, then we'll update this later

Reply:
I think it would be better to not _have_ to change it later - someone might see that "composer_args" is plural and therefore not even think to check that they need to change it if they add an additional arg.

Reply:
Updated